feat: improve multisig utility and usability

s8sato opened the pull request:
BREAKING CHANGES:

- (api-changes) `CanRegisterAnyTrigger` `CanUnregisterAnyTrigger` permissions for system authority
- (api-changes) `GenesisWasmTrigger` in `RawGenesisTransaction` for `genesis.json` readability
- (api-changes) `Multisig*Args` for multi-signature operations
- (config-changes) `genesis.json` assumes `wasm_triggers[*].action.executable` is prebuilt under `wasm_dir`
- (config-changes) docker-compose service generates `genesis.json` by oneself instead of reading bind-mounted one

Major commits:

- feat: support multisig recursion
- feat: introduce multisig quorum and weights
- feat: add multisig subcommand to client CLI
- feat: introduce multisig transaction time-to-live
- feat: predefine multisig world-level trigger in genesis
- feat: allow accounts in domain to register multisig accounts

## Context

<!-- - Describe the objective or issue this PR addresses, as well as the impact of the change. -->
<!-- - Try to keep the description accessible to newcomers. -->
<!-- - If you're resolving a specific issue, add "Fixes #issue_number" or "Closes #issue_number". -->

- Partially resolves #4930

Opens:

- #5022: essential to security
- #5185: essential to true utility
- #5160
- #5178

### Solution

<!-- - Describe the approach taken to achieve the objective / resolve the issue. -->

Each commit is explained below, starting with the most recent. You can see the commit history [here](https://github.com/s8sato/iroha/commits/wip/4930/)

#### feat: support multisig recursion

<!-- マルチシグが階層的に機能するようにする
組織の承認フローの自動化などに活用が期待される
マルチシグが提案されると末端の個人署名者に至るまですべての傘下のノード（MSアカウント）で承認を受け付ける準備が自動的に調う
CLIを用いて末端の個人署名者は直属のMSAへ提案されたものだけでなく自分が関与するすべてのMSTを参照できる -->

Allows multisig to function hierarchically, expected to be useful for e.g. automating organizational approval flows.

- When a multisig transaction proposed, every subordinate node (multisig account and corresponding transactions registry) automatically get ready to accept approvals that will be coming up from individual signatories
- Using CLI, individual signatories can query all pending multisig transactions they are involved in, not only those proposed to their direct multisig account

Tests:

```bash
cargo test -p iroha integration::multisig::multisig_recursion
bash scripts/tests/multisig.recursion.sh
```

#### feat: introduce multisig quorum and weights

[Sui]: https://docs.sui.io/concepts/cryptography/transaction-auth/multisig

<!-- SUIのマルチシグにインスパイアされた
「MofN」を超えて、完全に自由ではないにしても融通のきく認証ポリシーの定義を可能にする
たとえばQと同等のWは管理者権限を表す -->

Inspired by [Sui]'s multisig. Allows for flexible, if not completely free, authentication policies beyond "m of n". For example, weight equivalent to quorum represents administrative privileges

#### feat: add multisig subcommand to client CLI

```bash
$ cargo build
$ ./target/debug/iroha multisig

The subcommand related to multisig accounts and transactions

Usage: iroha multisig <COMMAND>

Commands:
  register  Register a multisig account
  propose   Propose a multisig transaction
  approve   Approve a multisig transaction
  list      List pending multisig transactions relevant to you
  help      Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help
```

You can see more usage in [the testing script](https://github.com/hyperledger/iroha/pull/5027/files#diff-e45d2b77c102ada9d021f1fd3e89a22d07a9ffa94d8fd56f1f3db725da702d02)

#### feat: introduce multisig transaction time-to-live

<!-- 最新のブロックのタイムスタンプを現在時刻と見なして時間切れの判定を行う
時間切れの判定はレジストリが呼ばれたときになされる -->

Considers the latest block timestamp as the current time and determines timeout, when the transactions registry is called

#### feat: predefine multisig world-level trigger in genesis

<!-- 全ドメインに権限を行使する世界レベルのトリガーをGで定義する
マルチシグに関連してシステム側には3種のトリガーがあることになる
_domains initializer_ 世界
ドメイン作成イベントに応じてドメイン所有者権限のARを登録する
_accounts registry_ ドメイン
呼び出しに応じてドメイン内のMSAを対応するTRとともに登録する
_transactions registry_ アカウント
呼び出しに応じてMSトランザクションの提案と承認を受け付け、実行を担当する -->

Defines a global trigger in genesis that exercises authority over all domains. There will be three types of triggers on the system side related to multisig:

- _Domains initializer_ has world-level authority and in response to a domain creation event, registers an accounts registry handled by the domain owner authority
- _Accounts registry_ has domain-level authority and when called by a domain resident, registers a multisig account along with a  transactions registry
- _Transactions registry_ has account-level authority and when called by a multisig signatory, accepts proposals or approvals for multisig transactions, and is responsible for executing them

#### feat: allow accounts in domain to register multisig accounts

<!-- Aレジストリはドメイン所有者相当の権限をもつので従来はアクセスが制限されていた
このコミットはドメイン内の限りにおいて誰でもどんなマルチシグでも組織できるようにする
これは緩め過ぎかもしれない
コード内にTODOとしてコメントしてある -->

Accounts registry has authority of the domain owner, so access was previously restricted. This commit allows anyone to organize any multisig account within the domain.

This may be too lenient. [Discussion](https://github.com/hyperledger/iroha/pull/5027#discussion_r1742108046)

---

### Review notes

<!-- 概要を把握するには -->

To get an overview,

```bash
cargo test -p iroha integration::multisig::multisig
bash scripts/tests/multisig.sh
```

```mermaid
sequenceDiagram
    autonumber
    participant oo as etc.
    participant DI as Domains Initializer
    Note over DI: /world
    oo-->>DI: domain created
    create participant AR as Accounts Registry
    DI-->>AR: register
    Note over AR: /world/domain
    create actor s0 as signatory 0
    oo->>s0: register
    create actor s1 as signatory 1
    oo->>s1: register
    s0->>AR: request new ms account
    create actor 01 as ms account 01
    AR-->>01: register
    create participant TR as Transactions Registry
    AR-->>TR: register
    AR-->>s0: grant ms role
    AR-->>s1: grant ms role
    Note over 01,TR: /world/domain/account
    s1->>TR: propose instructions
    create participant tx as pending ms transaction
    TR-->>tx: deploy ms transaction
    s0->>TR: approve instructions
    destroy tx
    TR-->>tx: execute instructions
```

The dotted line indicates an automatic process

<!-- - For complex PRs, try to provide some information on how to approach the review more effectively. -->
<!-- - For example, is there a natural order in which the affected files should be reviewed? -->

### Checklist

- [x] I've read [`CONTRIBUTING.md`](../CONTRIBUTING.md).
- [x] I've written integration tests for the code changes.
- [ ] All review comments have been resolved.

nxsaken commented:
Schema needs to be updated

s8sato commented on defaults/genesis.json:
![image](https://github.com/user-attachments/assets/169f7317-2693-48ed-90b5-e1b7623ac38c)

Oddly enough, genesis is up to date locally and passes [the consistency check](https://github.com/s8sato/iroha/blob/f5790c2048f800adcd34716cc44cc28810b78609/.github/workflows/iroha2-dev-pr.yml#L38). Maybe has something to do with embedding wasm

s8sato commented on client/tests/integration/multisig.rs:
```diff
@@ -1,77 +1,209 @@
+    )?;
 
-    let args = MultisigRegisterArgs {
+    // Create a multisig account ID and discard the corresponding private key
+    // FIXME #5022 Should not allow arbitrary IDs. Otherwise, after #4426 pre-registration account will be hijacked as a multisig account
+    let multisig_account_id = gen_account_in(kingdom).0;
```
I think the real solution is that a multisig account creation request derives a string that is a variant of multihash but without the authentication functionality, as suggested in #5022.

If all we want to do were to prevent hijacking, we can simply impose a signature on all or part of the payload of the request. However, this does not eliminate the risk that the requester may unfairly hold the private key

s8sato replied:
Left to https://github.com/hyperledger/iroha/issues/5022#issuecomment-2390938541

s8sato commented on wasm_samples/multisig_domains/src/lib.rs:
```diff
@@ -0,0 +1,48 @@
+        ),
+    );
+
+    Register::trigger(accounts_registry)
```
The former involves #5022, and the latter, when combined with #4670, also involves #5022

s8sato replied:
Yes, so regarding transactions registry and multisig role, that contain multihash, such collisions will rarely occur. But it is still problematic with accounts registry that contain only a domain name

s8sato replied:
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: avoid name collisions of multisig role and triggers

Erigara reviewed (commented):
I think this change is great improvement to usability of multisigs.

s8sato commented on client_cli/src/main.rs:
```diff
@@ -1197,6 +1200,227 @@ mod json {
+        pub signatories: Vec<AccountId>,
+        /// Relative weights of responsibility of respective signatories
+        #[arg(short, long, num_args(2..))]
+        pub weights: Vec<u8>,
```
If they find themselves totally powerless in the organization, they'll simply stop participating. There seems to be no particular reason to prohibit 0, just as there is no particular reason to prohibit 255

s8sato commented on client_cli/src/main.rs:
```diff
@@ -1197,6 +1200,227 @@ mod json {
+                .parse()
+                .unwrap();
+            let account = Account::new(account);
+            let signatories = signatories.into_iter().zip(weights).collect();
```
Yeah, the former seems better to me, because from the interface perspective it would be better to iterate over ids and numbers separately

s8sato replied:
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: signatories and weights must be equal in length

s8sato commented on docs/source/references/schema.json:
Yes, I'll try a similar way as the executor

mversic replied:
you may also put the initialization code into executor migration, but I think `genesis.json` is better. Sth like this:
```json
{
  "chain": "00000000-0000-0000-0000-000000000000",
  "executor": "./executor.wasm",
  "smart_contracts": [
    "./smart_contract_1",
    "./smart_contract_2"
  ]
}
```

s8sato replied:
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: `genesis.json` includes not wasm blobs but hashes

s8sato commented on scripts/tests/multisig.recursion.sh:
I'm thinking the same thing. It would have been faster to write it in python from the beginning

s8sato replied:
Left to https://github.com/hyperledger/iroha/issues/4930#issuecomment-2390929346

s8sato commented on tools/kagami/src/genesis/generate.rs:
```diff
@@ -127,34 +130,79 @@ pub fn generate_default(
+    // Allow the initializer to register a multisig accounts registry for any domain
+    let grant_to_register_any_trigger =
+        Grant::account_permission(CanRegisterAnyTrigger, multisig_world_authority);
+    // Manually register a multisig accounts registry for wonderland whose creation in genesis does not trigger the initializer
+    let register_multisig_accounts_registry_for_wonderland = {
```
Great suggestion, sounds like a domain for administrators.

Isn't that impossible to achieve in one round, regardless of the order in the genesis?

s8sato replied:
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: add a domain for domain-free system accounts

s8sato commented on wasm_samples/executor_custom_data_model/src/multisig.rs:
```diff
@@ -1,24 +1,32 @@
+    pub signatories: BTreeMap<AccountId, u8>,
+    /// Threshold of total weight at which the multisig is considered authenticated
+    pub quorum: u16,
+    /// Multisig transaction time-to-live based on block timestamps. Defaults to [`DEFAULT_MULTISIG_TTL_SECS`]
```
I guess you are talking about individual transactions

s8sato replied:
For now it is not allowed, but we could consider making the default indefinite

s8sato replied:
Left to https://github.com/hyperledger/iroha/issues/4930#issuecomment-2390929346

s8sato commented on wasm_samples/multisig_domains/src/lib.rs:
```diff
@@ -0,0 +1,48 @@
+        Action::new(
+            executable,
+            Repeats::Indefinitely,
+            // FIXME #5022 This trigger should continue to function regardless of domain ownership changes
+            domain.owned_by().clone(),
```
I didn't know the way to listen for multiple type of events. If that's possible, the initializer can switch behaviors depending on whether domain is created or transferred

Erigara replied:
I've outline how to do that on trigger side: [here](https://github.com/hyperledger/iroha/pull/5027#discussion_r1743469956)
And when registering trigger: [here](https://github.com/hyperledger/iroha/pull/5027#discussion_r1743469956)

s8sato replied:
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: perpetuate multisig domain-level authority

s8sato commented on wasm_samples/multisig_transactions/src/lib.rs:
```diff
@@ -0,0 +1,225 @@
+            .map(|(_, weight)| weight as u16)
+            .sum();
+
+    let is_expired = {
```
Yes, expiration is determined only when the transaction registry is called

Erigara commented on tools/kagami/src/genesis/generate.rs:
```diff
@@ -127,34 +130,79 @@ pub fn generate_default(
+                executable,
+                Repeats::Indefinitely,
+                multisig_world_authority.clone(),
+                DomainEventFilter::new().for_events(DomainEventSet::Created),
```
Event filter which can match multiple events:

```rust
DomainEventFilter::new().for_events(DomainEventSet::Created | DomainEventSet::OwnerChanged),
```

Erigara commented on wasm_samples/multisig_domains/src/lib.rs:
```diff
@@ -0,0 +1,48 @@
+
+#[iroha_trigger::main]
+fn main(_id: TriggerId, _owner: AccountId, event: EventBox) {
+    let domain = match event {
+        EventBox::Data(DataEvent::Domain(DomainEvent::Created(domain))) => domain,
+        _ => dbg_panic("should be triggered only by domain created events"),
+    };
```
So here you should match multiple events:

```rust
match event {
    EventBox::Data(DataEvent::Domain(DomainEvent::Created(domain))) => {
        // ...
    }
    EventBox::Data(DataEvent::Domain(DomainEvent::OwnerChanged(DomainOwnerChanged { domain, new_owner }))) => {
        // ...
    }
}
```

s8sato commented:
Updates:

- review doc: add a reference to multisig recursion diagram
- review fix: signatories and weights must be equal in length
- review fix: condition to rebuild multisig subordinate triggers
- review fix: `genesis.json` includes not wasm blobs but hashes
- review fix: autofill multisig account domains
- review fix: avoid name collisions of multisig role and triggers
- review fix: perpetuate multisig domain-level authority
- review fix: add a domain for domain-free system accounts

[commit history](https://github.com/s8sato/iroha/commits/wip/4930/)

Notes:

- Currently script tests (`scripts/tests/multisig.*`) are failing after rebase for some reason

s8sato commented on scripts/tests/multisig.recursion.sh:
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: add a reference to multisig recursion diagram

s8sato commented on wasm_samples/multisig_accounts/src/lib.rs:
```diff
@@ -0,0 +1,158 @@
+        ),
+    );
+
+    Register::trigger(multisig_transactions_registry)
```
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: avoid name collisions of multisig role and triggers

s8sato commented on wasm_samples/multisig_accounts/src/lib.rs:
```diff
@@ -0,0 +1,158 @@
+    let role =
+        Role::new(role_id.clone()).add_permission(can_execute_multisig_transactions_registry);
+
+    Register::role(role)
```
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: avoid name collisions of multisig role and triggers

s8sato commented on wasm_samples/multisig_accounts/build.rs:
```diff
@@ -0,0 +1,20 @@
+const TRIGGER_DIR: &str = "../multisig_transactions";
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    println!("cargo::rerun-if-changed={}", TRIGGER_DIR);
```
Resolved in https://github.com/hyperledger/iroha/pull/5027#issuecomment-2390722239: condition to rebuild multisig subordinate triggers

s8sato commented on wasm_samples/multisig_transactions/src/lib.rs:
```diff
@@ -0,0 +1,225 @@
+        MultisigTransactionArgs::Propose(instructions) => HashOf::new(instructions),
+        MultisigTransactionArgs::Approve(instructions_hash) => *instructions_hash,
+    };
+    let instructions_metadata_key: Name = format!("proposals/{instructions_hash}/instructions")
+        .parse()
+        .unwrap();
+    let proposed_at_ms_metadata_key: Name = format!("proposals/{instructions_hash}/proposed_at_ms")
+        .parse()
+        .unwrap();
+    let approvals_metadata_key: Name = format!("proposals/{instructions_hash}/approvals")
+        .parse()
+        .unwrap();
```
I'd resolve this as it is not necessary from multisig perspective. Discussions around the granular metadata access are seen in https://github.com/hyperledger/iroha/issues/4353#issuecomment-2146573078, #4782, #5023

s8sato commented on client/tests/integration/multisig.rs:
```diff
@@ -1,146 +1,405 @@
+    test_client.submit_blocking(Log::new(Level::DEBUG, "Just ticking time".to_string()))?;
+
+    // Check that the multisig accounts registry has been generated
+    let multisig_accounts_registry_id = multisig_accounts_registry_of(&kingdom.parse().unwrap());
```
Left to https://github.com/hyperledger/iroha/issues/4930#issuecomment-2390929346

s8sato commented on client/tests/integration/multisig.rs:
```diff
@@ -1,77 +1,209 @@
+
+    // Any account in the same domain can register a multisig account without special permission
+    let not_signatory_client = alt_client(not_signatory, &test_client);
+    not_signatory_client
```
Left to https://github.com/hyperledger/iroha/issues/4930#issuecomment-2390929346

s8sato commented on hooks/pre-commit.sample:
```diff
@@ -3,15 +3,14 @@
 # update the default executor
 cargo run --release --bin iroha_wasm_builder -- build ./wasm_samples/default_executor --optimize --out-file ./defaults/executor.wasm
 # update the default genesis, assuming the transaction authority is `iroha_test_samples::SAMPLE_GENESIS_ACCOUNT_ID`
+mkdir -p defaults/blobs
+rm defaults/blobs/*
 cargo run --release --bin kagami -- genesis generate --executor-path-in-genesis ./executor.wasm --genesis-public-key ed01204164BF554923ECE1FD412D241036D863A6AE430476C898248B8237D77534CFC4 > ./defaults/genesis.json
```
After this PR merge, you will locally need to re-generate `genesis.json` as needed to get correct wasm blobs referenced as hashes in `genesis.json`

s8sato replied:
When [approach changed](https://github.com/hyperledger/iroha/pull/5027#discussion_r1787878806), you will be running another command introduced by #5123 to get wasm blobs

s8sato commented on defaults/genesis.json:
```diff
@@ -149,6 +166,75 @@
+        }
+      }
+    },
+    {
+      "Register": {
+        "Trigger": {
+          "id": "multisig_accounts_wonderland",
+          "action": {
+            "executable": {
+              "Wasm": "6f94f34b715a40aa0b98ea3a6e83107844e55de1625de6b6dbbfdec125a1ad07"
+            },
+            "repeats": "Indefinitely",
+            "authority": "ed0120CE7FA46C9DCE7EA4B125E2E36BDB63EA33073E7590AC92816AE1E861B7048B03@wonderland",
+            "filter": {
+              "ExecuteTrigger": {
+                "trigger_id": "multisig_accounts_wonderland",
+                "authority": null
+              }
+            },
+            "metadata": {}
+          }
+        }
+      }
     }
```
I'd change this approach to fit #5123 since the wasm samples is superset of the genesis trigger blobs.

Instead of interpreting trigger registration as an ordinary instruction, we can give the genesis triggers a special field in `genesis.json` just like the executor, as like @mversic originally suggested:

```json
{
  "executor": "./executor.wasm",
  "wasm_triggers": [
    "multisig_accounts_wonderland": {
      "executable": "multisig_accounts",
      "repeats": "Indefinitely",
      "authority": "ed0120CE7FA46C9DCE7EA4B125E2E36BDB63EA33073E7590AC92816AE1E861B7048B03@wonderland",
      "filter": {
        "ExecuteTrigger": {
          "trigger_id": "multisig_accounts_wonderland",
          "authority": null
        }
      },
      "metadata": {}
    }
  ]
}
```

then `executable: &str` will be `WasmSmartContract` via `load_sample_wasm()` to compose a `Register<Trigger>` instruction.

Note that you will need to precompile wasm samples not only on integration tests, but also whenever genesis is needed

mversic commented on defaults/genesis.json:
```diff
@@ -149,6 +166,75 @@
+          "id": "multisig_domains",
+          "action": {
+            "executable": {
+              "Wasm": "cb253c4ebe8ae68ad6591b4dd7acc447282df663fc16a3cc20f9938bfa0d3d83"
```
didn't we say this should be path to the file?

s8sato replied:
This was an attempt to improve readability while minimize changes except for `executable.Wasm` representation, but yeah, I'm changing approach to specify a path under `wasm_samples`. See https://github.com/hyperledger/iroha/pull/5027#discussion_r1787878806

Erigara commented on crates/iroha_data_model/src/transaction.rs:
```diff
@@ -62,27 +64,20 @@ mod model {
     /// Wrapper for byte representation of [`Executable::Wasm`].
     ///
     /// Uses **base64** (de-)serialization format.
```
Docs should be adjusted to reflect that (de)serialization format is no longer always base64.

s8sato commented:
Updates:

- review fix: exception for multisig role registration
- review fix: give `genesis.json` a special field to register wasm triggers
- DROPPED: review fix: `genesis.json` includes not wasm blobs but hashes

[commit history](https://github.com/s8sato/iroha/commits/wip/4930/)

Notes:

- > Currently script tests (`scripts/tests/multisig.*`) are failing after rebase for some reason
  
  This was because only integration tests passed due to my oversight of [privileged Alice in `TestGenesis`](https://github.com/hyperledger/iroha/blob/242d857a333a497f8afa712b941e7184ae35e7b6/crates/iroha_test_network/src/lib.rs#L108). Fixed
- Stopped pursuing the cause of the failing `integration::extra_functional::restart_peer::restarted_peer_should_have_the_same_asset_amount`

mversic commented on crates/iroha_genesis/src/lib.rs:
```diff
@@ -356,6 +389,53 @@ impl Decode for ExecutorPath {
+#[derive(Debug, Clone, Serialize, Deserialize, IntoSchema, Encode, Decode)]
+pub struct GenesisWasmAction {
+    /// Expected to be converted by [`iroha_test_samples::load_sample_wasm`]
+    executable: String,
```
```suggestion
    executable: PathBuf,
```

like `ExecutorPath`

s8sato replied:
Changed `ExecutorPath` to `WasmPath` and shared it

mversic commented on crates/iroha_genesis/src/lib.rs:
```diff
@@ -39,12 +41,13 @@ pub struct RawGenesisTransaction {
     instructions: Vec<InstructionBox>,
-    /// Initial topology
-    topology: Vec<PeerId>,
+    wasm_triggers: Vec<GenesisWasmTrigger>,
```
I'd also prefer to put it after `instructions` so that they can execute triggers but it's just not possible atm because triggers depend on some instructions executing before them

mversic commented on crates/iroha_executor/src/default.rs:
```diff
@@ -1270,6 +1307,28 @@ pub mod trigger {
+        let trigger_name = trigger.id().name().as_ref();
+        let naming_is_ok = if let Some(tail) = trigger_name.strip_prefix("multisig_accounts_") {
+            let multisig_system: AccountId =
+                // iroha_test_samples::MULTISIG_SYSTEM_ID
+                "ed01201F677E0900C2F633391310D12D155112DF65EDF9DC800D13797CEE5DAF47B890@system"
+                    .parse()
+                    .unwrap();
```
if you want to have hardcoded `system` domain and account, then they shouldn't be defined in the `genesis.json`. Rather they should be always implicitly created

mversic replied:
> Like genesis domain and account, right?

yes

s8sato replied:
Coded them in `iroha_genesis::GenesisBuilder::default`, not the pre-genesis initial state

mversic commented on wasm_samples/executor_custom_data_model/src/multisig.rs:
```diff
@@ -1,24 +1,32 @@
+    /// Threshold of total weight at which the multisig is considered authenticated
+    pub quorum: u16,
+    /// Multisig transaction time-to-live based on block timestamps. Defaults to [`DEFAULT_MULTISIG_TTL_SECS`]
+    pub transaction_ttl_secs: Option<u32>,
```
```suggestion
    #[serde(default = DEFAULT_MULTISIG_TTL_SECS)]
    pub transaction_ttl_ms: u32,
```

1. can we do it like this?
2. we use `ms` in `transaction.rs` as unix timestamp

I think there should always be a TTL defined. User an just put a huge number

s8sato replied:
Ok, we can accept seconds as an user input and convert it to milliseconds here

s8sato commented on crates/iroha_data_model/src/block.rs:
```diff
@@ -459,9 +459,9 @@ mod candidate {
+            if transactions.len() > 5 {
                 return Err(
-                    "Genesis block must have 1 to 4 transactions (executor upgrade, initial topology, parameters, other isi)",
+                    "Genesis block must have 1 to 5 transactions (executor upgrade, initial topology, parameters, other instructions, trigger registrations)",
```
https://github.com/hyperledger/iroha/pull/5027#discussion_r1793243148

0x009922 commented on crates/iroha_cli/src/main.rs:
```diff
@@ -1197,6 +1200,232 @@ mod json {
+        pub quorum: u16,
+        /// Time-to-live of multisig transactions made by the multisig account
+        #[arg(short, long)]
+        pub transaction_ttl_secs: Option<u32>,
```
Let's be kind to users: https://docs.rs/humantime/latest/humantime/

mversic commented on wasm_samples/multisig_accounts/build.rs:
```diff
@@ -0,0 +1,27 @@
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    for path in [
+        TRIGGER_DIR,
+        "../../crates/iroha_data_model/src",
+        "../../crates/iroha_core/src/smartcontracts",
+        "../../crates/iroha_smart_contract/src",
```
I don't think this is needed. We could put this into all smart contracts then? @0x009922

0x009922 replied:
Yeah, this list of dependencies is not strictly exhaustive.

IMO it's better to run these build scripts unconditionally - `iroha_wasm_builder` relies on `cargo`, thus rebuilds only when necessary.

_However_, it's possible to extend the wasm builder to read [Cargo messages](https://docs.rs/cargo_metadata/0.18.1/cargo_metadata/enum.Message.html), track all involved dependencies, and print them with `cargo::rerun-if-changed={path}`. Sounds like an overkill to me though.

s8sato replied:
Deleted the entire `multisig_*/build.rs`. Instead, made the script build them in sequence

mversic commented on wasm_samples/multisig_accounts/src/lib.rs:
```diff
@@ -0,0 +1,167 @@
+
+#[iroha_trigger::main]
+fn main(id: TriggerId, owner: AccountId, event: EventBox) {
+    let args: MultisigAccountArgs = match event {
+        EventBox::ExecuteTrigger(event) => event
+            .args()
+            .dbg_expect("args should be attached")
+            .try_into_any()
+            .dbg_expect("args should be for a multisig account"),
+        _ => dbg_panic("should be triggered by a call"),
+    };
```
```suggestion
    let EventBox::ExecuteTrigger(event) => event else {
        panic!("Unexpected event type");
    };
        
    let args: MultisigAccountArgs = event
        .args()
        .try_into_any()
        .dbg_expect("args should be for a multisig account");
```

mversic commented on wasm_samples/executor_custom_data_model/src/multisig.rs:
I noted this somewhere else. I don't think this belongs here mixed with all data model for tests. IMO you should make a separate crate called `multisig_data_model` just for this purpose

s8sato replied:
Created `iroha_multisig_data_model`. Also, separated `wasm_samples` into `wasm/libs/multisig_*` and `wasm/samples/*`

mversic commented on crates/iroha_genesis/src/lib.rs:
```diff
@@ -39,12 +41,13 @@ pub struct RawGenesisTransaction {
     /// Path to the [`Executor`] file
     executor: ExecutorPath,
+    /// Initial topology
+    topology: Vec<PeerId>,
```
> The idea is no gap with [the actual execution order](https://github.com/hyperledger/iroha/blob/90940a1816f4f68d458c0a657c26f709e1726c8b/crates/iroha_genesis/src/lib.rs#L154).

could you change the execution order?

mversic replied:
created #5158

mversic commented on crates/iroha_cli/src/main.rs:
```diff
@@ -1197,6 +1200,232 @@ mod json {
+        pub account: AccountId,
+        /// Instructions to approve
+        #[arg(short, long)]
+        pub instructions_hash: iroha::crypto::Hash,
```
regarding this, I noticed that `HashOf` implements `Deserialize`. IMO then it could also implement `FromStr`. What do you think?

mversic commented on crates/iroha_cli/Cargo.toml:
```diff
@@ -30,6 +30,7 @@ path = "src/main.rs"
 iroha = { workspace = true }
 iroha_primitives = { workspace = true }
 iroha_config_base = { workspace = true }
+executor_custom_data_model = { version = "=2.0.0-rc.1.0", path = "../../wasm_samples/executor_custom_data_model" }
```
just apply [this](https://github.com/hyperledger/iroha/pull/5027#discussion_r1805944134) suggestion

mversic commented on crates/iroha_cli/src/main.rs:
```diff
@@ -1197,6 +1200,232 @@ mod json {
+                instructions_hash,
+            } = self;
+            let registry_id: TriggerId = format!(
+                "multisig_transactions_{}_{}",
```
> Isn't the scoping feasible by permission of the trigger authority?

but it's not real scoping because we can't have 2 triggers of the same name. For instance in your case if we could have `multisig_transactions` scoped under `account` you could have multiple instances of the same trigger under different accounts. Instead you have to rely on parsing trigger id

mversic replied:
I've seen this pattern many times. I think this is something to be considered outside the scope of this PR

mversic replied:
opened #5160

mversic commented on crates/iroha_cli/src/main.rs:
```diff
@@ -1197,6 +1200,232 @@ mod json {
+                .as_ref()
+                .strip_prefix("multisig_signatory_")
+                .unwrap()
+                .replace('_', "@")
```
let's come up with a separation character sequence that we should use when naming triggers. Something that's unlikely to be used by clients. How about `_%_`?

s8sato replied:
Resolved by changing to `.replacen('_', "@", 1)` atm. We can assume a multihash doesn't contain punctuations

0x009922 commented on crates/iroha/tests/integration/multisig.rs:
```diff
@@ -77,76 +218,275 @@ fn mutlisig() -> Result<()> {
+    let account_in_another_domain = gen_account_in("garden_of_live_flowers").0;
+
+    {
+        let reserved_prefix = "multisig_accounts_";
```
I would make the prefix more "special" like `__MULTISIG_ACCOUNT__`

s8sato replied:
I think this suggestion can be connected to the scoping discussion #5160

0x009922 commented on wasm_samples/Cargo.toml:
```diff
@@ -23,8 +23,9 @@ members = [
     "smart_contract_can_filter_queries",
-    "multisig_register",
-    "multisig",
+    "multisig_accounts",
+    "multisig_domains",
+    "multisig_transactions",
```
I wonder if it is really necessary to have MS functionality to be spread across three smartcontracts? I wish all the logic to be contained within a single one. To support different "modes" of operation (i.e. as a domain initialiser, accounts registry, transactions registry) it can rely on various metadata.

s8sato replied:
For clarity, they are hierarchical according to their authorities atm

0x009922 commented on crates/iroha_genesis/src/lib.rs:
```diff
@@ -356,6 +389,53 @@ impl Decode for ExecutorPath {
+
+impl From<GenesisWasmAction> for Action {
+    fn from(src: GenesisWasmAction) -> Self {
+        let executable = load_sample_wasm(src.executable);
```
All paths in configuration are relative to the source file they are specified at.

Please make sure that this path is resolved relatively to the path of `genesis.json` itself.

0x009922 commented on crates/iroha_genesis/src/lib.rs:
```diff
@@ -39,12 +41,13 @@ pub struct RawGenesisTransaction {
     instructions: Vec<InstructionBox>,
-    /// Initial topology
-    topology: Vec<PeerId>,
+    wasm_triggers: Vec<GenesisWasmTrigger>,
```
It seems to introduce inconsistency to me - why should triggers registration be different from other instructions? What if I need to intermix them in a custom order? etc.

I guess the answer is that it's hard to achieve due to the data model limitations - since we deserialise them as `InstructionBox`, we need to somehow patch the internals to load trigger from path. I think it is a bad path.

However, it's not the only way. It's possible to read `genesis.json` as a raw `JsonValue` and pre-process it before parsing into a data-model type. I.e. traverse `instructions` field, find `Register`/`Trigger`, find `executable_path` field there, load WASM, put it as a WASM blob into in-memory JSON value, then deserialise as a valid `InstructionBox`.

(same could be done for `executor` btw - load, put as a plain `Upgrade` instruction)

This might be an overkill for this PR, however. But from the standpoint of UX and API consistency this seems to be a good thing to do.

s8sato replied:
- Created #5178

0x009922 commented on crates/iroha_cli/src/main.rs:
```diff
@@ -1197,6 +1200,232 @@ mod json {
     }
 }
+
+mod multisig {
+    use std::io::{BufReader, Read as _};
+
+    use executor_custom_data_model::multisig::{MultisigAccountArgs, MultisigTransactionArgs};
```
I don't immediately like the idea of having this being included directly into the client. Though, it depends on how much we need to make `multisig` an official out-of-the-box feature.

As an alternative idea, there could a separate CLI to construct multisig-related instructions in a form of JSON, which _then_ could be supplied to the client:

```shell
iroha_multisig register --account .. --weights .. --quorum .. | iroha json transaction
```

s8sato replied:
This can be discussed together with the client interface issues left to #4930

s8sato commented:
Major updates:

- accept `transaction_ttl` as `humantime::Duration` and convert to milliseconds
- hide system entities configuration from `kagami::genesis::generate`
- reorganize `wasm_samples`, remove `multisig_*/build.rs`
- create dedicated `iroha_multisig_data_model` crate

Minor updates are fixed up with existing commits.

[commit history](https://github.com/s8sato/iroha/commits/wip/4930/)

Notes:

- `genesis_block_builder_example` is failing to compile for some reason. Left to another PR

I'd address further feedbacks in another PR of #4930. This PR is already enough hard to maintain

mversic commented on wasm/samples/default_executor/.cargo/config.toml:
this file can be removed

mversic commented on wasm/samples/default_executor/src/lib.rs:
maybe this should be in `libs`?

mversic replied:
tbh we should probably have `default_executor` (in libs) and `test_executor` (in samples)

s8sato replied:
We could argue that the default executor is a sample because the actual executor construction is the user's responsibility, but we could also argue that it is a library because custom executors are recommended to be migrated after genesis and the default executor is essential to the standard functionality.

For the latter reason I placed `default_executor` in `wasm/libs/`

mversic commented on wasm/libs/multisig_domains/src/lib.rs:
```diff
@@ -0,0 +1,79 @@
+
+#[iroha_trigger::main]
+fn main(host: Iroha, context: Context) {
+    let (domain_id, domain_owner, owner_changed) = match context.event {
+        EventBox::Data(DataEvent::Domain(DomainEvent::Created(domain))) => {
+            (domain.id().clone(), domain.owned_by().clone(), false)
+        }
+        EventBox::Data(DataEvent::Domain(DomainEvent::OwnerChanged(owner_changed))) => (
+            owner_changed.domain().clone(),
+            owner_changed.new_owner().clone(),
+            true,
+        ),
+        _ => dbg_panic("should be triggered only when domain created or owner changed"),
+    };
+
```
```suggestion
    let EventBox::Data(DataEvent::Domain(event)) = context.event else {
        dbg_panic("Not a domain event");
    };

    let (domain_id, domain_owner, owner_changed) = match event {
        DomainEvent::Created(domain) => {
            (domain.id().clone(), domain.owned_by().clone(), false)
        }
        DomainEvent::OwnerChanged(owner_changed) => (
            owner_changed.domain().clone(),
            owner_changed.new_owner().clone(),
            true,
        ),
        _ => dbg_panic("should be triggered only when domain created or owner changed"),
    };

```

mversic commented on docs/source/references/schema.json:
I think you should add multisig data model into schema

mversic commented on crates/iroha_multisig_data_model/src/lib.rs:
```diff
@@ -0,0 +1,64 @@
+#[derive(Serialize, Deserialize)]
+pub struct MultisigAccountArgs {
+    /// Multisig account to be registered
+    /// WARNING: any corresponding private key allows the owner to manipulate this account as a ordinary personal account
```
```suggestion
    /// # Warning
    ///
    /// Any corresponding private key allows the owner to manipulate this account as a ordinary personal account
```

mversic commented on scripts/tests/multisig.sh:
```diff
@@ -0,0 +1,66 @@
+MULTISIG_ACCOUNT=$(gen_account_id "msa")
+WEIGHTS=($(yes 1 | head -n $N_SIGNATORIES)) # equal votes
+QUORUM=$N_SIGNATORIES # unanimous
+TRANSACTION_TTL="1y 6M 2w 3d 12h 30m 30s 500ms"
```
IMO this is pretty unusable to CLI users. I haven't seen CLI tools with such API.
I'd much rather you revert to using seconds

s8sato replied:
There seems to be a disagreement between reviewers. I don't have a strong opinion on this. May I leave this to #4930 @mversic  @0x009922  ?

0x009922 replied:
Yeah, I totally disagree with Marin on that. @nxsaken @DCNick3 what do you think?

Anyway, I am okay with postponing it and using any solution for now.

mversic commented on crates/iroha/Cargo.toml:
```diff
@@ -85,7 +85,8 @@ iroha_genesis = { workspace = true }
 iroha_test_network = { workspace = true }
 iroha_executor_data_model = { workspace = true }
-executor_custom_data_model = { version = "=2.0.0-rc.1.0", path = "../../wasm_samples/executor_custom_data_model" }
+iroha_multisig_data_model = { workspace = true }
```
we might want to make `multisig_data_model` a regular dependency of `iroha` and re-export it so that users can interact with multisig without additional complexity

mversic commented on crates/iroha_genesis/src/lib.rs:
```diff
@@ -423,6 +591,7 @@ mod tests {
         // First transaction
         {
             let transaction = transactions[0];
+            // FIXME field `value` of struct `CommittedTransaction` is private
             let instructions = transaction.value.instructions();
```
```suggestion
            let instructions = transaction.as_ref().instructions();
```

mversic commented on crates/iroha_genesis/src/lib.rs:
```diff
@@ -381,6 +548,7 @@ mod tests {
         Ok(())
     }
 
+    #[cfg(not_compile)]
```
```suggestion
```

s8sato reviewed (commented):
Updates:

- re-export as `iroha::multisig_data_model`
- add multisig data model into schema
- `default_executor` in `wasm/libs`
- early panic with unexpected events
- rustdoc-compliant warning
- empty arg to build all wasms
- fix `genesis_block_builder_example` test

All changes since the last update contained.

[commit history](https://github.com/s8sato/iroha/commits/wip/4930/)

Notes:

- > `genesis_block_builder_example` is failing to compile for some reason. Left to another PR

  Thanks to @mversic , fixed

s8sato commented on defaults/genesis.json:
![image](https://github.com/user-attachments/assets/0af0db2a-a76d-4e21-aed2-755256bbb547)

Wasm libraries need to have path resolution at genesis and upload/download in CI, as well as the executor @s8sato

s8sato commented:
Updates:

- fix CI to upload and download wasm libraries
- FIXUP: make a wasm trigger executable path relative to `defaults/genesis.json`
- FIXUP: hide system entities configuration from `kagami::genesis::generate`
- EDIT: review fix: give `genesis.json` a special field to register wasm triggers

[commit history](https://github.com/s8sato/iroha/commits/wip/4930/)

Notes:

![image](https://github.com/user-attachments/assets/84f994c0-d133-4b91-8319-c75afe359250)

- Does this have something to do with the repository change?

s8sato commented:
![image](https://github.com/user-attachments/assets/9aebb784-da72-4c13-b3a3-30edba18a9ac)

> This request was automatically failed because there were no enabled runners online to process the request for more than 1 days.

Every failing job `runs-on: [self-hosted, Linux, iroha2]`. It seems they depend on runners not available atm

s8sato commented:
Afterword: I think it was not a good practice to keep a single commit during review, which was remnants of our `Rebase and merge` days. Now that we `Squash and merge`, I'd separate commits to some extent
